src/main/java/com/stefancooper/SpigotUHC/events/BaseEvents.java

```diff
@@ -206,4 +226,196 @@ public void onChatMessage(AsyncPlayerChatEvent event) {
             event.setMessage(String.format("(Death Chat) %s", event.getMessage()));
         }
     }
-}
+
+    @EventHandler
+    public void onInventoryClick(InventoryClickEvent event) {
+        if (!(event.getWhoClicked() instanceof Player player)) return;
+
+        // Handle all inventory click types (shift-click, drag-drop, etc.)
+        config.getManagedResources().runTaskLater(() -> {
+            final ItemStack helmet = player.getInventory().getHelmet();
+
+            if (AdditionalEnchants.isNightVisionGoggles(helmet)) {
+                if (!player.hasPotionEffect(PotionEffectType.NIGHT_VISION)) {
+                    player.addPotionEffect(new PotionEffect(
+                            PotionEffectType.NIGHT_VISION,
+                            Integer.MAX_VALUE,
+                            0,
+                            true,
+                            false
+                    ));
+                }
+            } else {
+                if (player.hasPotionEffect(PotionEffectType.NIGHT_VISION)) {
+                    player.removePotionEffect(PotionEffectType.NIGHT_VISION);
+                }
+            }
+        }, 1L); // Delay by 1 tick so inventory updates first
+    }
+
+    @EventHandler
+    public void onPlayerInteract(PlayerInteractEvent event) {
+        if (event.getHand() != EquipmentSlot.HAND) return; // ignore off-hand
+        if (event.getItem() == null || event.getAction() != Action.RIGHT_CLICK_AIR && event.getAction() != Action.RIGHT_CLICK_BLOCK)
+            return;
+
+        final Player player = event.getPlayer();
+        final ItemStack item = event.getItem();
+
+        // Delay to allow helmet slot to update
+        config.getManagedResources().runTaskLater(() -> {
+            final ItemStack helmet = player.getInventory().getHelmet();
+
+            if (AdditionalEnchants.isNightVisionGoggles(helmet)) {
+                if (!player.hasPotionEffect(PotionEffectType.NIGHT_VISION)) {
+                    player.addPotionEffect(new PotionEffect(
+                            PotionEffectType.NIGHT_VISION,
+                            Integer.MAX_VALUE,
+                            0,
+                            true,
+                            false
+                    ));
+                }
+            } else {
+                player.removePotionEffect(PotionEffectType.NIGHT_VISION);
+            }
+        }, 1L);
+    }
+
+    @EventHandler
+    public void onInventoryDrag(InventoryDragEvent event) {
+        if (!(event.getWhoClicked() instanceof Player player)) return;

```

Comment:
you can set final at this point too :) 
```suggestion
        if (!(event.getWhoClicked() instanceof final Player player)) return;

```